### PR TITLE
Refactor to dependency injection + unit tests

### PR DESCRIPTION
**Pull request by SIMULATAN**
# More unit tests
### Checklist

- [X] I have checked the PRs for upcoming features/bug fixes.
- [X] I have read the [contributing guidelines](/CONTRIBUTING.md).
- [X] I agree with the [license](/LICENSE).

### Changes
- [X] Refactoring
- [X] Internal code
- [ ] External Code / Libraries
- [ ] Documentation
- [ ] Other: \_____

Closes Issue: #172 

## Description
This adds more unit tests and refactors the internal code structure with proper design patterns.

## Todo
- [x] Update branch
- [x] Fix merge issues

## NOTE: Only squash merge this!

### CODE REVIEW

File: src/main/java/com/github/black0nion/blackonionbot/api/impl/get/Stats.java [hunk 1]
```diff
@@ -34,11 +33,11 @@ public Object handle(Context ctx, JSONObject body, Map<String, String> headers,
 					.put("cpu_speed", CPU_MHZ))
 			.put("guild_count", StatisticsManager.getGuildCount())
 			.put("user_count", StatisticsManager.getUserCount())
-			.put("prefix", Config.prefix)
+			.put("prefix", Config.getInstance().getPrefix())
 			.put("os", OS_NAME)
```
Comment (RealYusufIsmail) [inline src/main/java/com/github/black0nion/blackonionbot/api/impl/get/Stats.java:38]: After the slash command pr is merged you will have a lot of issues.

File: src/main/java/com/github/black0nion/blackonionbot/api/impl/post/Logout.java [hunk 2]
```diff
@@ -28,15 +23,10 @@ public class Logout implements IPostRoute {
...
-
-		if (doc != null) {
-			BlackSession.collection.deleteOne(Filters.eq("sessionid", ctx.header("sessionid")));
-			return "";
-		}
-		throw new UnauthorizedResponse("Invalid session id");
+	public Object handle(Context ctx, JSONObject body, Map<String, String> headers, RestSession session, DiscordUser user) throws Exception {
+		session.logout();
+		ctx.status(204);
+		return "";
```
Comment (RealYusufIsmail) [inline src/main/java/com/github/black0nion/blackonionbot/api/impl/post/Logout.java:29]: Add some Java doc

File: src/main/java/com/github/black0nion/blackonionbot/api/impl/post/Vote.java [hunk 3]
```diff
@@ -30,14 +30,14 @@ public class Vote implements IPostRoute {
...
 		assert ip != null;
-		if (!ip.equals("159.203.105.187") || !headers.get("authorization").equals(Config.topgg_auth)) {
+		if (!ip.equals("159.203.105.187") || !headers.get("authorization").equals(Config.getInstance().getTopggAuth())) {
 			throw new UnauthorizedResponse();
 		}
 
-		final long channelId = Config.vote_channel;
+		final long channelId = Config.getInstance().getVoteChannel();
 		if (channelId != -1) {
 			final TextChannel channel = Bot.getInstance().getJda().getTextChannelById(channelId);
```
Comment (RealYusufIsmail) [inline src/main/java/com/github/black0nion/blackonionbot/api/impl/post/Vote.java:42]: General thing but change veranda to getJDA

**Review by RealYusufIsmail — CHANGES_REQUESTED** [review]

**Comment by RealYusufIsmail** [comment]
Nice. I can now make sure my code is good.

**Comment by RealYusufIsmail** [comment]
I started adding the new issue format. Can you finish it off please @

File: src/main/java/com/github/black0nion/blackonionbot/api/impl/get/Stats.java (hunk 1 above)
Comment (SIMULATAN) [reply src/main/java/com/github/black0nion/blackonionbot/api/impl/get/Stats.java:37]: Aparently it worked well enough

File: src/main/java/com/github/black0nion/blackonionbot/api/sessions/RestSession.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.github.black0nion.blackonionbot.api.sessions;
+
+import java.util.InputMismatchException;
+import java.util.concurrent.ExecutionException;
+
+public non-sealed class RestSession extends GenericSession {
```
Comment (RealYusufIsmail) [inline src/main/java/com/github/black0nion/blackonionbot/api/sessions/RestSession.java:6]: Why dont use okhttp

File: src/main/java/com/github/black0nion/blackonionbot/mongodb/MongoDB.java [hunk 5]
```diff
@@ -5,6 +5,20 @@
 import com.mongodb.client.MongoDatabase;
 
 public class MongoDB {
-	public static final MongoDatabase DATABASE = MongoManager.getDatabase("BlackOnion-Bot" + (Config.run_mode != RunMode.RELEASE ? "_" + Config.run_mode.name() : ""));
-	public static final MongoDatabase generalDatabase = MongoManager.getDatabase("General");
+
```
Comment (RealYusufIsmail) [inline src/main/java/com/github/black0nion/blackonionbot/mongodb/MongoDB.java:8]: remove this space.

File: src/main/java/com/github/black0nion/blackonionbot/api/sessions/WebSocketSession.java [hunk 6]
```diff
@@ -9,9 +9,9 @@
 import java.util.InputMismatchException;
 import java.util.concurrent.ExecutionException;
 
-public class BlackWebsocketSession extends BlackSession implements org.eclipse.jetty.websocket.api.Session {
+public non-sealed class WebSocketSession extends GenericSession implements Session {
```
Comment (RealYusufIsmail) [inline src/main/java/com/github/black0nion/blackonionbot/api/sessions/WebSocketSession.java:12]: Why not use https://github.com/TakahikoKawasaki/nv-websocket-client

**Review by RealYusufIsmail — CHANGES_REQUESTED** [review]
Otherwise LGTM!

File: src/main/java/com/github/black0nion/blackonionbot/api/sessions/RestSession.java (hunk 4 above)
Comment (SIMULATAN) [reply src/main/java/com/github/black0nion/blackonionbot/api/sessions/RestSession.java:6]: It's used for the server side sessions

File: src/main/java/com/github/black0nion/blackonionbot/api/sessions/WebSocketSession.java (hunk 6 above)
Comment (SIMULATAN) [reply src/main/java/com/github/black0nion/blackonionbot/api/sessions/WebSocketSession.java:12]: Again, this is for the server side of things and it's just an utility class for logic like discord sessino handling. Thanks for reminding me that I have to write docs!

Comment (RealYusufIsmail) [reply src/main/java/com/github/black0nion/blackonionbot/api/sessions/WebSocketSession.java:12]: no problem/

**Review by RealYusufIsmail — APPROVED** [review]

**Review by RealYusufIsmail — APPROVED** [review]
Test it but all looks good.

**Comment by RealYusufIsmail** [comment]
A lot of classes to review

**Comment by SIMULATAN** [comment]
> A lot of classes to review 

Just point out what you find, just tell me if something sticks out. A few found issues is more than nothing :)

**Pull request merged**